src/Framework/IEventSource.cs

```diff
@@ -171,158 +172,44 @@ public interface IEventSource
     public static class EventSourceExtensions
     {
         /// <summary>
-        /// Helper method ensuring single deduplicated subscription to the <see cref="IEventSource.MessageRaised"/> event.
+        /// Helper method ensuring single deduplicated subscription to the event.
         /// </summary>
-        /// <param name="eventSource"></param>
+        /// <param name="removeHandler"></param>
+        /// <param name="addHandler"></param>
         /// <param name="handler">Handler to the event. If this handler is already subscribed, single subscription will be ensured.</param>
-        public static void HandleMessageRaised(this IEventSource eventSource, BuildMessageEventHandler handler)
+        private static void EnsureSingleSubscription<TEventHandler>(Action<TEventHandler> removeHandler, Action<TEventHandler> addHandler, TEventHandler handler)
+            where TEventHandler : Delegate
         {
-            eventSource.MessageRaised -= handler;
-            eventSource.MessageRaised += handler;
+            removeHandler(handler);
+            addHandler(handler);
         }
 
-        /// <summary>
-        /// Helper method ensuring single deduplicated subscription to the <see cref="IEventSource.ErrorRaised"/> event.
-        /// </summary>
-        /// <param name="eventSource"></param>
-        /// <param name="handler">Handler to the event. If this handler is already subscribed, single subscription will be ensured.</param>
-        public static void HandleErrorRaised(this IEventSource eventSource, BuildErrorEventHandler handler)
-        {
-            eventSource.ErrorRaised -= handler;
-            eventSource.ErrorRaised += handler;
-        }
+        public static void HandleMessageRaised(this IEventSource eventSource, BuildMessageEventHandler handler) => EnsureSingleSubscription(h => eventSource.MessageRaised -= h, h => eventSource.MessageRaised += h,  handler);
```

Comment:
Open to opinion, but this refactor actually feels less clean to me. It adds a lot of delegate overhead, and the saved lines are almost all either from comments or from taking multiple lines and putting them end-to-end. Furthermore, EnsureSingleSubscription doesn't actually ensure anything; it just calls the two methods in a preset order. If I were to call EnsureSingleSubscription(h => eventSource.ErrorRaised += h, h => eventSource.ErrorRaised -= h, handler), for instance, it would compile fine but do the wrong thing.

Reply:
The idea was to shorten a code duplication, because of the listed methods do same and repetitive thing.

Let's hear others.

Reply:
I still feel it's an improvement - as it more explicitly expresses the intent of the code (ensuring single subscription).

In the code size and typo risk it's same (not decreasing nor increasing). So as a grant total it feels as improvement

Reply:
Though much leaner but not clear. Also, error prone as mentioned by @Forgind!